hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/SparkBaseIndexSupport.scala

```diff
@@ -155,32 +158,59 @@ abstract class SparkBaseIndexSupport(spark: SparkSession,
       (List.empty, List.empty)
     } else {
       var recordKeyQueries: List[Expression] = List.empty
-      var recordKeys: List[String] = List.empty
-      for (query <- queryFilters) {
-        val recordKeyOpt = getRecordKeyConfig
-        RecordLevelIndexSupport.filterQueryWithRecordKey(query, recordKeyOpt).foreach({
-          case (exp: Expression, recKeys: List[String]) =>
-            recordKeys = recordKeys ++ recKeys
-            recordKeyQueries = recordKeyQueries :+ exp
-        })
+      var compositeRecordKeys: List[String] = List.empty
+      val recordKeyOpt = getRecordKeyConfig
+      val isComplexRecordKey = recordKeyOpt.map(recordKeys => recordKeys.length).getOrElse(0) > 1
+      recordKeyOpt.foreach { recordKeysArray =>
+        // Handle composite record keys
+        breakable {
+          for (recordKey <- recordKeysArray) {
+            var recordKeys: List[String] = List.empty
+            for (query <- queryFilters) {
+              {
+                RecordLevelIndexSupport.filterQueryWithRecordKey(query, Option.apply(recordKey), isComplexRecordKey).foreach {
+                  case (exp: Expression, recKeys: List[String]) =>
+                    exp match {
+                      // For IN, add each element individually to recordKeys
+                      case _: In => recordKeys = recordKeys ++ recKeys
+
+                      // For other cases, basically EqualTo, concatenate recKeys with the default separator
+                      case _ => recordKeys = recordKeys ++ recKeys
+                    }
+                    recordKeyQueries = recordKeyQueries :+ exp
+                }
+              }
+            }
+
+            if (recordKeys.isEmpty) {
+              recordKeyQueries = List.empty
+              compositeRecordKeys = List.empty
+              break()
+            } else if (!isComplexRecordKey || compositeRecordKeys.isEmpty) {
+              compositeRecordKeys = recordKeys
+            } else {
+              var tempCompositeRecordKeys: List[String] = List.empty
+              for (compRecKey <- compositeRecordKeys) {
```

Comment:
This is for "Cartesian product" for m * n pattern: where m and n are enumerations of the possible values for a record key field.

Reply:
Still we need to check all the fields of the composition records keys are involved in the query predicates, for e.g. if we have `col_0,col_1,col_2` as the record key fields, then the query predicate `col_0 = val_0 and col_1 = val_1` does not work? Only `col_0 = val_0 and col_1 = val_1 and col_2 = val_2` should work if we do not matches the MDT RLI with prefix search of the keys.

Also we should be very careful with the field sequence in the returned `compositeRecordKeys`, if `col_0,col_1,col_2` are the record key fields, then literal `col_0:val_0,col_2:val_2,col_1:val_1` does not work either.

Reply:
Added test cases. These cases are handled.